tests/test_errors.py

```diff
@@ -196,7 +196,7 @@ def finalize(self, cfg: Any) -> None:
         Expected(
             create=lambda: OmegaConf.create({"foo": "${missing}"}),
             op=lambda cfg: getattr(cfg, "foo"),
-            exception_type=ConfigKeyError,
+            exception_type=ConfigAttributeError,
```

Comment:
I wonder if missing interpolations should have their own exception? (that would neither be a KeyError nor an AttributeError)

Reason is that people may use exception-based programming with the new API, to detect missing keys:

```python
try:
    foo = cfg.foo
except AttributeError:
    foo = cfg.foo = 42
print(foo)
```
and if the `AttributeError` is raised because `foo` is referring to a missing interpolation, then bad things may happen.

Reply:
Why is that bad?
you accessed a field, it was pointing to something that is not there. you assign an actual value to it. It's no longer a pointer but a real value. underlying pointed to field is unchanged.

Reply:
Someone writing such code most likely only intended to override `cfg.foo` if there was no such key in the config already. This could lead to subtle bugs if suddenly an interpolation gets broken (e.g., because a variable was renamed in the config), because instead of seeing an error the code will run just fine (ignoring the broken interpolation).

Reply:
Create a concrete example demonstrating a problem.
writing values on an interpolation is not writing through the interpolation.
Also, your example will not even work now because the exception raised is not an ConfigAttributeError but a ConfigKeyError right now:

```
cfg = OmegaConf.create({"a" : "${b}"})
cfg.a # ConfigKeyError
cfg["a"] # ConfigKeyError
```
and in any case writing on an interpolation is not write through:
```python
﻿In [11]: cfg
Out[11]: {'a': '${b}'}
In [12]: cfg.a = 10
In [13]: cfg
Out[13]: {'a': 10}
```
I am open to changes here but it's a different topic (and does not much much to do with standardizing with dict and object (given that they do not support interpolations).

Reply:
Are you on master? I'm at d018678 and in your example `cfg.a` raises a `ConfigAttributeError` for me.

Here's an example illustrating what I have in mind: note how the typo in the interpolation makes it being silently ignored.

```python
cfg = OmegaConf.create("""
default_ram: 32
super_computer_ram: 128

machines: [m1, m2, m3, m4, m5]

ram_per_machine:
    m3: 64
    m5: ${super_computerr_ram}
""")

# Ensure `cfg.ram_per_machine` contains RAM data for all machines.
for machine in cfg.machines:
    ram_per_machine = cfg.ram_per_machine
    try:
        ram = getattr(ram_per_machine, machine)
    except AttributeError:
        ram = cfg.default_ram
        setattr(ram_per_machine, machine, cfg.default_ram)
    print(f"RAM for {machine}: {ram}")

print("\nConfig:\n")
print(OmegaConf.to_yaml(cfg, resolve=True))
```

It's not a huge deal, I just thought I'd point it out (it wasn't a concern before because there were no exceptions to catch)

Reply:
> Are you on master? 

nope, I was not. Exceptions are correct.

Looking at your example, I am not sure what you are trying to say.
1. What is the expected output?
2. Is the typo in the config intentional?

If the typo (`computerr`) is intentional, I don't think I care.
Your config is wrong, you are applying custom logic and it does not do what you wanted because your config is wrong.

Reply:
>     1. What is the expected output?

If you fix the typo in the config, the output is:
```
RAM for m1: 32
RAM for m2: 32
RAM for m3: 64
RAM for m4: 32
RAM for m5: 128
```
>     2. Is the typo in the config intentional?

I think the idea is that the typo was not intentional. The `AttributeError` is being interpreted by this custom logic as meaning that `ram_per_computer.m5` was not set, but the reality is that `ram_per_computer.m5` *was* set, but there was an accidental typo in the `ram_per_computer.m5` setting (preventing the interpolation from being resolved).

Reply:
> I wonder if missing interpolations should have their own exception? (that would neither be a KeyError nor an AttributeError)

So the idea is that each of the following would raise a different type of exception:
```
OmegaConf.create("").a  # raises ConfigAttributeError
OmegaConf.create("a: ${xyz}").a  # also raises ConfigAttributeError
```

Reply:
@Jasha10 said it better than I could have (thank you!)

Again, I don't feel strongly about it, since it should be pretty rare (combination of exception-based programming + interpolation + user error) => just wanted to mention the idea of maybe using a different exeption, but up to you @omry

Reply:
We already have
```class InterpolationResolutionError(OmegaConfBaseException, ValueError): ...```
defined in `omegaconf.errors`. Why no throw that one?

Reply:
Ha, looks like I added it when I added [relative interpolations](https://github.com/omry/omegaconf/commit/3c208f948da2569c475baba1aff70338ece8210b) but somehow never used it.

I am fine with using it to indicate an that an interpolation could not be resolved.

Reply:
I've opened a PR (https://github.com/omry/omegaconf/pull/523) to use `InterpolationResolutionError` for this purpose.